doc/design/draft/named-segments.md

```diff
@@ -0,0 +1,228 @@
+
+## Design
+
+### Updated RouDi Config
+
+Supporting named segments will require add an additional `name` field to the RouDi config under the `[[segment]]` heading, as well as updating the config version Example:
+
+```
+[general]
+version = 2
+
+[[segment]]
+name = "foo"
+
+[[segment.mempool]]
+size = 32
+count = 10000
+```
+
+Updating the version will be necessary for older versions of RouDi to fail informatively when presented with a newer config. Otherwise RouDi can support version 1 configs as if they are version 2 configs with no `name` fields specified.
+
+### Requesting Segments
+
+#### When creating a Publisher
+
+A new field will be added to the [PublisherOptions struct](../../../iceoryx_posh/include/iceoryx_posh/popo/publisher_options.hpp) as follows:
+
+```
+struct PublisherOptions
+{
+  ...
+  ShmName_t segmentName{""};
+```
+
+#### When creating a Subscriber
+
+A similar field will be added to the [SubscriberOptions struct](../../../iceoryx_posh/include/iceoryx_posh/popo/subscriber_options.hpp), except that it will support multiple elements
+
+```
+struct SubscriberOptions
+{
+  ...
+  vector<ShmName_t, MAX_SUBSCRIBER_SEGMENTS> segmentNames{};
+```
+
+`MAX_SUBSCRIBER_SEGMENTS` can be set to some reasonably small value to start with since explicitly allowing many but not all read-access segments is an unlikely use case. 
+
+If this assumption turns out to be wrong however, we can always update it to be `MAX_SHM_SEGMENTS` instead.
+
+#### When creating Clients and Servers
+
+Clients and Servers will have similar fields, distinguished by request/response:
+
+```
+struct ClientOptions
+{
+  ...
+  ShmName_t requestSegmentName{""};
+  vector<ShmName_t, MAX_RESPONSE_SEGMENTS> responseSegmentNames{};
+```
+
+```
+struct ServerOptions
+{
+  ...
+  vector<ShmName_t, MAX_REQUEST_SEGMENTS> requestSegmentNames{};
+  ShmName_t responseSegmentName{""};
+```
+
+#### When initializing the runtime
+
+We may wish to also additionally support requesting segments upon runtime initialization. The motivation here is that if we know we will only need to use certain segments, we can avoid expending unecessary resources mapping the ones we will not use. Additionally, this can be used as an error mitigation technique to prevent communication endpoints from requesting an unsupported segment.
+
+Currently initialization of the runtime happens in the constructor, and any errors that occur result in a contract violation and program termination. 
+
+While it is not directly in scope of this design, it would be beneficial to refactor the Runtime implementations to use the builder pattern such that any errors may be handled through returning errors types.
+
+Tying this together with requesting segments, this could look something like:
+```
+class DefaultRuntimeBuilder
+{
+  ...
+  IOX_BUILDER_PARAMETER(vector<ShmName_t>, shmSegmentNames, {});
+  IOX_BUILDER_PARAMETER(bool, allowUnmappedSegments, false);
```

Comment:
Although the boolean would also work in this case, in general I prefer enums since they are more future proof and have more than two options. It could be something like
```cpp
enum class UnmappedSegmentBehavior {
    Ignore,
    LoadOnDemand,
    TerminateProcess,
    ReleaseSampleAndReturnError
};
```

`UnmappedSegmentBehavior::Ignore` could be the default and allows to create a subscriber and afterwards calling `runtime.mapReadOnlySegments`. Even if nothing is called, it might be interesting when the application is not interested in the data itself but might only forward the sample.

Reply:
Good point. An enum would work fine. Note that this value only applies to create of Publishers/Clients/Servers when they request a segment. I haven't considered the possibility of detecting we have a sample from an unregistered segment and suspect it would be non-trivial to add the ability in this scenario to map the segment on demand (the smart chunk data type almost certainly doesn't have access to the segment manager).

Reply:
I think we need to add some hooks to the building blocks the be able to gracefully handle this situation. We should be able to get the information from the pointer repository and then inform someone that we cannot read the payload. This should be possible without reading the chunk header. Just dropping the sample and printing a warning would also be an option. After all, not being able to read the payload is most probably just a misconfiguration.

Reply:
Yeah I agree. `take` should probably return an informative error, at the very least. 

Looking into the code, it is unclear to me so far where exactly things will fail. At a high level I know that the subscriber queue contains offset pointers. I don't know exactly if/when that pointer might be dereferenced and crash because the segment is not mapped. But presumably this happens somewhere during the `take()` and can be guarded against, propagating an error out. 

Then the question is whether we want to just return the error to the user, or whether we want to allow some sort of hook in the subscriber to ask the runtime to map that segment.

Reply:
So... I THINK that all the relative pointer stuff and pointer repository stuff will just return nullptrs when referencing an unmapped segment id. Reasoning about taking and releasing chunks in the subscriber in this sense is okay. It can increment and decrement the reference in the management segment no problem, without actually having access to the chunk data itself. 

The first potential problem I see is at this line of code: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl#L57

It looks like the subscriber gets a chunk, doesn't actually check if the chunk is a nullptr, and then plows into dereferencing the nullptr with:
```
        auto receivedChunkHeaderVersion = chunk.getChunkHeader()->chunkHeaderVersion();
```

I'd have to do some testing to confirm my suspicion. This is already a bug unto its own right. Outside of this design we should have better error handling for this scenario.